protocol/forge-scripts-deploy-in-go.md

```diff
@@ -0,0 +1,222 @@
+# Forge Scripts Deploy in Go
+
+# Purpose
+
+The purpose of this doc is to improve the integration between the `forge script` based chain deployment/genesis flow
+and the Go tooling and testing.
+
+# Summary
+
+By running the chain-deployment/genesis solidity scripts in an instrumented Go EVM we can reduce the roundtrip time,
+and greatly improve integration with inputs (configs, dependencies) and outputs (state, artifacts).
+
+This unlocks runtime-customization of deployments in Go tests,
+to increase coverage and create new multi-L2 chain deployments for Interop tests.
+
+# Problem Statement + Context
+
+From [Design doc 52](https://github.com/ethereum-optimism/design-docs/pull/52):
+
+> The current L2 chain deployment approach originates from a time with Hardhat,
+> single L1 target, and a single monolithic set of features.
+> 
+> Since then the system has migrated to Foundry and extended for more features,
+> but remains centered around a single monolithic deploy-config for all its features.
+> 
+> **With Interop we need to configure a new multi-L2 deployment**:
+> the number of ways to compose L2s in tests grows past what a single legacy config template can support.
+> 
+> Outside of interop, deployment also seems increasingly complex and opaque, while it does not have to be,
+> due to the same configuration and composability troubles.
+
+See the design-doc for further in-depth context on the problem.
+
+The integration between (1) Go testing/tooling and (2) Forge deployment/genesis needs to improve
+to reduce the complexity of deploying and testing.
+
+Specifically, Go tools need reliable and organized inputs for deployment/genesis,
+and interface with the forge scripts artifacts or scripts themselves
+in a way that is configurable but not error-prone.
+
+# Proposed Solution
+
+In [deployment-chains design doc 52](https://github.com/ethereum-optimism/design-docs/pull/52) and
+[OPStackManager design doc 60](https://github.com/ethereum-optimism/design-docs/pull/60) the importance
+of Modular configs and incremental deploy steps is outlined.
+
+To utilize those modular configs and deploy steps, we need a way to either 
+run and cache the individual Forge script functions (as outlined in 52),
+or integrate it closer such that no caching complexity is required.
+
+This document proposes the latter: run the Forge scripts within Go,
+such that we do not need multiple steps of Forge script sub-process to build a state for testing / genesis tooling.
+
+## Running Forge scripts in Go
```

Comment:
I have slight concern about possible bugs in the foundry Go precompile implementation resulting in e2e tests not running against "the same" contracts. I suppose the precompile isn't super complex, but adding layers of abstraction does make me feel worried